### DISCUSSION TITLE
Towards a low cost harp-board for ecosystem dissemination

### DISCUSSION THREAD

**Discussion by bruno-f-cruz** [harp-tech/protocol#108] [Open-ended]
# A low-cost harp board for ecosystem dissemination

## Goals

- Create a low-cost harp board with basic functionality
- Provide a competitive alternative to run Bonsai courses with instead of Arduino
- Provide a cheap "first harp board" for people to get started with harp and try out the ecosystem
- Provide a demo board that can be used to play with custom firmware and build new devices

## Features

I think a minimal viable product includes:

### Hardware

- 4 digital inputs
- 4 digital outputs
- (these pins could potentially be configurable between input and output)
- 1 PWM cable output (frequencies compatible with servos would be a plus)
- 1 Analog input (250 Hz sampling rate would suffice)
- 1 harp-compatible synch input (for syncing with other harp boards)
- 1 USB port for power and communication
- 1 LED for harp status

### Firmware

- Implement core harp protocol + harp synch protocol
- USB communication
- Control over digital output lines (eg. Set, Clear, Toggle, State)
- Asynchronous events on digital input line state changes (i.e. Rising and Falling)
- Configurable PWM output (frequency, duty cycle) and a register to start and stop it
- A periodic event with an analog input read

## Implementation

I think the obvious choice for this board is the pico. Without going into too much detail, I think the pico is a good choice because:

- An existing harp core implementation is available
- It's cheap
- Easy to source
- It's performant
- It's easy to program (at the limit, just click a button and drag and drop a uf2)
- Lots of IOs
- Good community support
- Great documentation
- An ecosystem with available interface boards

### Hardware

At the level of the hardware I think the most viable strategy is to try to use something that already exist in the market to interface with outputs. Discussing currently available options, we came up with the following options:

- Gravity
    https://www.dfrobot.com/product-2393.html

- Grove
    https://www.seeedstudio.com/Grove-Shield-for-Pi-Pico-v1-0-p-4846.html

The obvious problem is that these boards do not include a physical layer for the clock synchronization input. To solve that, @poofjunior suggested that we could adopt a "shim board" strategy that sandwiches between the pico and the shield board. [Something like this.](https://www.robotshop.com/products/motor-shim-for-pico?srsltid=AfmBOopY47YPTJESJkBlKKdDb5B9yn0W0uWGdNwZVR7ZXwlElA_pL1Oento). This would obviously co-opt a couple of pins from the pico that may be available to the shield board, but I think it's a reasonable trade-off.

Given the prices of the individual parts, I believe we could probably offer a complete solution for <20$.

### Firmware

Not a lot to say here just yet. But I will just signal that this seems like a good opportunity to write a small tutorial on "How to generate your own device using the pico harp core"

**Comment by jfrazao** [comment]
I would prefer to have more analog and pwm/servo eventually reducing the number of digital io.
Raspberry pico can do:
3 ADC
3 PWM

**Comment by bruno-f-cruz** [comment]
For a tentative first course using this board:

## Exercises

## DAY1

### Setup 
- Install Bonsai (Link to the install guide)
- Install Harp.HobGoblin 
- Flash Pico with .uf2
- "Hello world" - like thing. Wire the button to the Analog Input. Device -> Parse(Analog). Make sure there is a workflow to copy paste. 
- install uv/harp python

### Acquisition and tracking

- Start with pico

  - Saving AnalogData (Lightsensor) and show how to do it with CSV (with headers).
   - load csv with normal pandas
 - Digital Output (LED) how to save with CSV (with header)
   - load csv with pandas (align with analog input)
 - DeviceDataWriter to save all registers
    - use harp-python to open and plot
 - Camera + videowriter
   - camera + sum region + csv writer 
 - Trigger camera with pulse train. @jfrazao to test if this is doable.
   - Wire another digital output to camera
   - Put the LED next to the photodiode + visible to camera 

(Optional Servo) for people that move quickly

## Day2

  - Reaction time task with digital input and digital output
     (Add harp timestamping + OS timestamping and compare jitters)   
  - Add timeout
  - Make it a go/no go
  - Experiment with different variants
 
## Day 3

###Closed-loop experiment

Track an animal with camera
Introduce tracking in bonsai (tracking pipeline)
Stimulate when tracking goes to a specific place (Task logic)
Use Harp to blink LED

Analyze data
